aten/src/ATen/native/TensorIndexing.cpp

```diff
@@ -154,5 +39,174 @@ std::ostream& operator<<(std::ostream& stream, const std::vector<TensorIndex>& t
   return stream;
 }
 
+// NOTE: Why do we mirror instead of replace the `count_specified_dimensions` function
+// in torch/csrc/autograd/python_variable_indexing.cpp? It's because
+// `count_specified_dimensions` is on the hot path of Python tensor multi-dim indexing
+// (i.e. it's called by `applySlicing` which is called by `THPVariable_getitem` /
+// `THPVariable_setitem` when handling indexing of more than one dimension). If we were
+// to merge the Python/C++ `count_specified_dimensions` function, on the Python side
+// we would have to construct a `std::vector` container to be consumed by the C++
+// `count_specified_dimensions` function, which adds 100s of nanoseconds overhead and
+// is undesirable.
+int64_t count_specified_dimensions(ArrayRef<TensorIndex> indices) {
+  // Count the number of indexed dimensions (everything but ellipsis and None)
+  int64_t count = 0;
+  size_t size = indices.size();
+  for (auto& obj : indices) {
+    if (obj.is_tensor()) {
+      auto& tensor = obj.tensor();
+      if (tensor.scalar_type() == kByte || tensor.scalar_type() == kBool) {
+        count += tensor.dim();
+      } else {
+        count++;
+      }
+    } else if (!obj.is_none() && !obj.is_ellipsis() && !obj.is_boolean()) {
+      count++;
+    }
+  }
+  return count;
+}
+
+// This mirrors `applySlicing` in torch/csrc/autograd/python_variable_indexing.cpp
+Tensor applySlicing(
+    const Tensor& self,
+    ArrayRef<TensorIndex> indices,
+    std::vector<Tensor>& outIndices,
+    const at::Device& self_device,
+    const IntArrayRef& self_sizes) {
+  int64_t dim = 0;
+  int64_t specified_dims = count_specified_dimensions(indices);
+
+  TORCH_CHECK_INDEX(specified_dims <= self_sizes.size(), "too many indices for tensor of dimension ", (int)self_sizes.size());
+
```

Comment:
Worth noting that you don't fastpath single element case (unlike the Python version) because you assume that TensorIndex is conveniently laid out in list form already?